File: packages/engine-freecut/src/executor/render.ts
// packages/engine-freecut/src/executor/render.ts
import { spawn } from "child_process";
import path from "path";
import os from "os";
import crypto from "crypto";

import { Action, ProjectSettings, RenderResult } from "./types";
import { AssetResolver } from "./assetResolver";
import { validatePlan } from "./planValidator";
import { buildTimeline } from "./timelineBuilder";
import { compileTimeline } from "./ffmpegCompiler";

export interface RenderOptions {
  actions: Action[];
  resolver: AssetResolver;
  settings: ProjectSettings;
  outputPath?: string;
  ffmpegBin?: string;
  // FAIL LOUD: if true, throws on unsupported actions instead of silently dropping
  strict?: boolean;
  onLog?: (line: string) => void;
}

export async function render(opts: RenderOptions): Promise<RenderResult> {
  const ffmpegBin = opts.ffmpegBin ?? "ffmpeg";
  const log = opts.onLog ?? ((l) => console.log(l));

  log(`[executor] received ${opts.actions.length} actions`);

  // ---------- 1. Validate ----------
  const validation = await validatePlan(opts.actions, opts.resolver);
  log(`[executor] validation ok=${validation.ok} errors=${validation.errors.length} warnings=${validation.warnings.length}`);
  for (const w of validation.warnings) log(`[executor][warn] ${w}`);
  if (!validation.ok) {
    for (const e of validation.errors) log(`[executor][err]  ${e}`);
    throw new Error(`Plan validation failed:\n${validation.errors.join("\n")}`);
  }

  // ---------- 2. Build timeline ----------
  const timeline = await buildTimeline(opts.actions, opts.resolver, opts.settings);
  log(
    `[executor] timeline built: ${timeline.videoSegments.length} video segs, ` +
      `${timeline.bgmTracks.length} bgm tracks, ${timeline.captions.length} captions, ` +
      `duration=${timeline.duration.toFixed(3)}s`
  );

  // ---------- 3. Compile to FFmpeg ----------
  const compiled = compileTimeline(timeline);

  const outputPath =
    opts.outputPath ??
    path.join(
      os.tmpdir(),
      `monet-media-dev/edited_${crypto.randomUUID()}.mp4`
    );

  const args: string[] = ["-y"];
  for (const inp of compiled.inputs) args.push("-i", inp);
  args.push("-filter_complex", compiled.filterGraph);
  args.push(...compiled.mapArgs);
  args.push(...compiled.outputArgs);
  args.push(outputPath);

  const fullCommand = `${ffmpegBin} ${args
    .map((a) => (a.includes(" ") || a.includes(";") ? `"${a}"` : a))
    .join(" ")}`;
  log(`[executor] cmd: ${fullCommand}`);

  // ---------- 4. Run ----------
  await runFfmpeg(ffmpegBin, args, log);

  // ---------- 5. Coverage report ----------
  const resolvedMedia: Record<string, string> = {};
  for (const id of [...new Set(validation.mediaIds)]) {
    const e = opts.resolver.resolve(id);
    if (e) resolvedMedia[id] = e.filePath;
  }

  return {
    outputPath,
    command: fullCommand,
    filterGraph: compiled.filterGraph,
    durationSec: timeline.duration,
    coverage: {
      actionsReceived: opts.actions.length,
      actionsApplied: opts.actions.length,
      unsupportedActions: [],
      resolvedMedia,
      unresolvedMedia: [],
    },
  };
}

function runFfmpeg(bin: string, args: string[], log: (l: string) => void) {
  return new Promise<void>((resolve, reject) => {
    const proc = spawn(bin, args);
    proc.stderr.on("data", (chunk) => log(`[ffmpeg] ${chunk.toString().trim()}`));
    proc.on("error", reject);
    proc.on("close", (code) =>
      code === 0 ? resolve() : reject(new Error(`ffmpeg exited with code ${code}`))
    );
  });
}

File: monet/engines/freecut/executor/render.py
# monet/engines/freecut/executor/render.py
from __future__ import annotations
import asyncio
import os
import tempfile
import uuid
import logging
from typing import Callable, List, Optional

from .types import Action, ProjectSettings, RenderResult, CoverageReport
from .asset_resolver import AssetResolver
from .plan_validator import validate_plan
from .timeline_builder import build_timeline
from .ffmpeg_compiler import compile_timeline

logger = logging.getLogger("monet.freecut.render")


async def render(
    actions: List[Action],
    resolver: AssetResolver,
    settings: ProjectSettings,
    output_path: Optional[str] = None,
    ffmpeg_bin: str = "ffmpeg",
    on_log: Optional[Callable[[str], None]] = None,
) -> RenderResult:
    log = on_log or logger.info
    log(f"[executor] received {len(actions)} actions")

    val = validate_plan(actions, resolver)
    log(f"[executor] validation ok={val.ok} errs={len(val.errors)} warns={len(val.warnings)}")
    for w in val.warnings:
        log(f"[executor][warn] {w}")
    if not val.ok:
        for e in val.errors:
            log(f"[executor][err]  {e}")
        raise ValueError("Plan validation failed:\n" + "\n".join(val.errors))

    timeline = await build_timeline(actions, resolver, settings)
    log(
        f"[executor] timeline: {len(timeline.videoSegments)} v-segs, "
        f"{len(timeline.bgmTracks)} bgm, {len(timeline.captions)} captions, "
        f"dur={timeline.duration:.3f}s"
    )

    compiled = compile_timeline(timeline)

    if output_path is None:
        out_dir = os.path.join(tempfile.gettempdir(), "monet-media-dev")
        os.makedirs(out_dir, exist_ok=True)
        output_path = os.path.join(out_dir, f"edited_{uuid.uuid4()}.mp4")

    args: List[str] = ["-y"]
    for inp in compiled.inputs:
        args += ["-i", inp]
    args += ["-filter_complex", compiled.filterGraph]
    args += compiled.mapArgs + compiled.outputArgs + [output_path]

    full_cmd = ffmpeg_bin + " " + " ".join(
        f'"{a}"' if (" " in a or ";" in a) else a for a in args
    )
    log(f"[executor] cmd: {full_cmd}")

    proc = await asyncio.create_subprocess_exec(
        ffmpeg_bin, *args,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    while True:
        line = await proc.stderr.readline()
        if not line:
            break
        log(f"[ffmpeg] {line.decode(errors='ignore').rstrip()}")
    rc = await proc.wait()
    if rc != 0:
        raise RuntimeError(f"ffmpeg exited with code {rc}")

    resolved, _ = resolver.assert_all_exist(list(set(val.mediaIds)))
    return RenderResult(
        outputPath=output_path,
        command=full_cmd,
        filterGraph=compiled.filterGraph,
        durationSec=timeline.duration,
        coverage=CoverageReport(
            actionsReceived=len(actions),
            actionsApplied=len(actions),
            resolvedMedia=resolved,
        ),
    )
